extmod/vfs_lfsx.c: reformat a path name into canonical form

robert-hh opened the pull request:
It sorts out . and .. in the path name and removes multiple /'s
Question I have doubts about:
- It uses m_malloc() to allocate space for the new path name from
  the heap. How do I prevent gc_collect() from freeing it, before it
  is used. Not likely, but I did not follow all possible cases.
- I have a temporary buffer for pointers which I've set to 32 entries.
  This is smaller than the worst case. It there something like vstring
  for other data types? Obviously I could use m_malloc and gc_free, or
  malloc and free, since this data is only used within the function.

robert-hh committed:
extmod/vfs_lfsx.c: reformat a path name into canonical form

It sorts out . and .. in the path name and removes multiple /'s
Question I have doubts about:
- It uses m_malloc() to allocate space for the new path name from
  the heap. How do I prevent gc_collect() from freeing it, before it
  is used. Not likely, but I did not follow all possible cases.
- I have a temporary buffer for pointers which I've set to 32 entries.
  This is smaller than the worst case. It there something like vstring
  for other data types? Obviously I could use m_malloc and gc_free, or
  malloc and free, since this data is only used within the function.

tve reviewed (commented):
I think something like this is screaming for a unit test...

I looked around a bit for how this is done and bumped into the algorithm described at https://en.cppreference.com/w/cpp/filesystem/path and I'm wondering whether a less cpu efficient but shorter version of this could be written inspired by that algorithm using something like:
```
// Eliminate duplicate slashes
while ((found=strstr(path, "//")) != null)
    memmove(found, found+1, strlen(found+1)+1);
// Replace "/./" by "/"
while ((found=strstr(path, "/./")) != null)
    memmove(found, found+2, strlen(found+2)+1);
// Eliminate trailing "/."
if (strcmp(path+strlen(path)-2, "/.") == 0)
    path[strlen(path)-2] = 0;
// Replace "/foo/../" by "/"
while (((found=strstr(path, "/../")) != null) || (strcmp(path+strlen(path)-3, "/..") == 0) {
    if (found == NULL) found = path+strlen(path)-3;
    *found = 0;
    if ((slash=strrchr(path, '/')) != null)
        memmove(slash, found+3, strlen(found+3)+1);
    else
        path = found+3; // may need to distinguish "foo/../etc" from "/../etc"?
}
```
I'm probably missing another corner case...

tve commented on extmod/vfs_lfsx.c:
```diff
@@ -98,6 +98,68 @@ STATIC void MP_VFS_LFSx(init_config)(MP_OBJ_VFS_LFSx * self, mp_obj_t bdev, size
+        if (token[n] == NULL) {
+            break;
+        }
+        in = NULL;
```
shouldn't this be `in = n`?

robert-hh replied:
No. in has to be a pointer; to the input string on the first call, NULL on all following ones.

tve replied:
Oops, I was not familiar with strtok, should have read the manual! What a weird interface...

robert-hh commented:
Thanks. After having sent that PR, I considered again and I'll make a new attempt as as a re-port of the little Python code below I made a few years ago. Obviously it is powered by the Python engine, which makes it short. Implementing in C would still require some allocation.
```
    def get_absolute_path(self, cwd, payload):
        # Just a few special cases "..", "." and ""
        # If payload start's with /, set cwd to /
        # and consider the remainder a relative path
        if payload.startswith('/'):
            cwd = "/"
        for token in payload.split("/"):
            if token == '..':
                cwd = self.split_path(cwd)[0]
            elif token != '.' and token != '':
                if cwd == '/':
                    cwd += token
                else:
                    cwd = cwd + '/' + token
        return cwd
```

robert-hh commented on extmod/vfs_lfsx.c:
```diff
@@ -98,6 +98,68 @@ STATIC void MP_VFS_LFSx(init_config)(MP_OBJ_VFS_LFSx * self, mp_obj_t bdev, size
 }
 
+// canonalize path, remove ., .. and // items
+static char *trim_path(const char *path) {
```
Yes, that's a good name.

robert-hh commented:
OK. so this is the C rewrite. Looks much better, but make heavy use of string routines. It works so far, but I'll integrate that into make_path, avoiding the clash with cur_dir, which is used as a temporary buffer. Until now tested separately.
```
static char *normalize_path(char *cwd, char *path_in)
{
    char *new_path = m_malloc(strlen(cwd) + strlen(path_in) + 1); // worst case
    char *path = strdup(path_in); // strdup allocates, so path must be free'd
    char *in = path;

    // Initiliaze new path
    // If cwd was empty (unlikely) or path starts with /, set it to /
    if (path[0] == '/' || cwd[0] == '\0') {
        strcpy(new_path, "/");
    } else {
        strcpy(new_path, cwd);
    }
    while (true) {  // leave it with a break
        char *token;
        token = strtok(in, "/");
        if (token == NULL) { // end of the path scan
            break;
        } 
        if (strcmp(token, "..") == 0) { // double slash, backup new_path
            char *p = strrchr(new_path, '/'); // should always work
            if (p) { // just for being sure
                p[p == new_path ? 1 : 0] = '\0'; // cut off the tail, but not the head
            }
        } else if (strcmp(token, ".") != 0) {
            if (strcmp(new_path, "/") != 0) { // not at the start
                strcat(new_path, "/");
            }
            strcat(new_path, token);
        }
        in = NULL;
    }
    free(path); // strdup allocates!
    return new_path;
}
```

tve commented:
Looks much nicer!
Dunno what to think about dynamically allocating new_path. I wonder whether it would make sense to pass new_path in as a parameter? Then it could perhaps go onto the stack. I haven't looked at call sites. That still leaves the strdup... Can you not clobber the incoming path?

robert-hh commented:
So I updated the PR with the new code.
No I cannot clobber the incoming path. strdup modifies it by inserting null characters.

tve commented:
Yup. Unless there's a reasonably small max path length that is enforced I'm not so sure that allocating on the stack is that hot...

robert-hh commented:
AFAIK, MAX_PATH is ~255.

robert-hh commented:
I changed the code to raise OSError(ENOMEM) if strdup fails. m_malloc raises internally an error if required. Looking through the varios places where make_path is used, I've seen some room for improvement:
- rename: the old_path is normalized, the new_path not. Applying the same call will simplify the code.
- remove and rmdir are exactly the same. Maybe one function can serve for both. They seem to be linked through a table, so the change would com at no costs.
- chdir: could be slightly simplified, since the result from make_path is always a new string.
-

robert-hh committed:
extmod/vfs_lfs(x).c: replace strtok with non-modifying code

At the cost of code size allocation for the temporary path is
avoided.
Also: make_path() does not return a const char* any more, so
declare it appropriately.

robert-hh committed:
extmod/vfs_lfsx.c: free the memory allocated by make_path

Even if that would be done by gc_alloc, early release seems fine.

robert-hh commented:
I replaced strtok by an equivalent function which does not modify the input string. That way,
strdup and free is not needed any more. But it increases the code, since strtok was already used
at another place. Since make_path does not return a const char * any more , I changed the declaration and added gc_free in the second commit as an early release of the memory.

robert-hh commented:
I assumed that the Null character is effectively added once.

davehylands commented:
strlen doesn't include the null character.

So if cwd was equal to "/home/foo" (strlen = 9) and path_in was equal to "a/b" (strlen = 3) then your result could be "/home/foo/a/b" (strlen = 13). You need to allocate 14 characters (9 + 1 for slash + 3 + 1 for terminating null).

robert-hh commented:
Right. I forgot the slash.

robert-hh commented:
Hi Dave, actually you are not right, but the reason is somewhat hidden, and I did not consider that initially. So yes, I forgot the slash, and no, it does not matter. because:
- if cwd was '/', then no '/' is added
- if cwd was not '/', then the string in cur_dir has a '/' at the end, which is removed before another path element is added. Removing makes backing up for '..' easier.

So it was right, by luck and not by intention.

robert-hh committed:
extmod/vfs_lfsx.c: Rewrite make_path because of travis fails

Quite a few platforms do not have strrchr, strncat or strncpy in
their libs. Verified by a local build attempt.
So make_path() is changed not to use these on the costs of readability.
Note: the omissions are strange:
esp8266: GENERIC_512K built, GENERIC and GENERIC_1M did not
stm32: PYBD_SF6 built, PYBD_SF2 and PYBV11 did not

tve commented on extmod/vfs_lfsx.c:
```diff
@@ -98,17 +99,68 @@ STATIC void MP_VFS_LFSx(init_config)(MP_OBJ_VFS_LFSx * self, mp_obj_t bdev, size
+            }
+            new_path[len_path] = '\0';
         }
+        token = get_token(NULL, '/', &len_token);
```
you could move this into the for statement...

robert-hh replied:
I moved that in and out. Stylewise I prefer to have it in, even if the generated code should not differ. Moving that into the loop body was is an attempt for silence a travis build fail, which still says in 13410.2:
```
In file included from ../../extmod/vfs_lfs.c:120:
../../extmod/vfs_lfsx.c:146:58: error: variable 'token' used in loop condition not modified in loop body [-Werror,-Wfor-loop-analysis]
    for (char *token = get_token(path, '/', &len_token); token != NULL;) {
```
All left-over travis complaints are now about re-definition of the function get_token, which is strange. 
```
In file included from ../../extmod/vfs_lfs.c:120:0:
../../extmod/vfs_lfsx.c:104:14: error: redefinition of ‘get_token’
 static char *get_token(const char *path, char sep, size_t *len) {
              ^
In file included from ../../extmod/vfs_lfs.c:74:0:
../../extmod/vfs_lfsx.c:104:14: note: previous definition of ‘get_token’ was here
 static char *get_token(const char *path, char sep, size_t *len) {
              ^
../../py/mkrules.mk:63: recipe for target 'build-coverage/extmod/vfs_lfs.o' failed
```

tve replied:
Looking at the errors, I believe that the "token error" is a side-effect of the get_token redefinition error, so once you solve that the token error will go away.

robert-hh replied:
But there is no re-definition, or at least I see none. get_token is defined only once.

tve replied:
Well, if you read the error messages carefully, it says `In file included from ../../extmod/vfs_lfs.c:120` and `In file included from ../../extmod/vfs_lfs.c:74`, so lines 74 and 120 both include `../../extmod/vfs_lfsx.c` and that's why there's a redefinition. (I haven't looked at the source.)

robert-hh replied:
But get_token is in the included vfs_lfsx.c, and therefore the definition is the same. I could declare it in vfs_lfs.c again, like it is done for make_path. Would that resolve the travis complaints?

tve replied:
Look at vfs_lfs.c carefully. It includes vfs_lfsx.c twice, once for LFS1 and once for LFS2. It defines some macros differently for each time so the functions end up being different and have different names. That's why there's the weird MP_VFS_LFSx(make_new) type of declarations. You need to move get_token out of the vfs_lfsx.c file.

robert-hh replied:
But get_token does not use any of these macros.

robert-hh replied:
Changed, and now it passes.

davehylands commented:
@robert-hh ok - the new code looks good as far as the allocation goes. I did find a bug in it though. I put together a gist over here: https://gist.github.com/davehylands/202ea7e2f1a80bad7ed03049d6a26b23

If you compile that and run it then you should get this output:
```
cur_dir =  path_in = a/b/c make_path = /a/b/c
cur_dir = / path_in = a/b/c make_path = /a/b/c
cur_dir = /home/ path_in = a/b/c make_path = /home/a/b/c
cur_dir =  path_in = /a/b/c make_path = /a/b/c
cur_dir = / path_in = /a/b/c make_path = /a/b/c
cur_dir = /home/ path_in = /a/b/c make_path = /a/b/c
cur_dir =  path_in = a/../c make_path = c
cur_dir = / path_in = a/../c make_path = c
cur_dir = /home/ path_in = a/../c make_path = /home/c
cur_dir =  path_in = /a/../c make_path = c
cur_dir = / path_in = /a/../c make_path = c
cur_dir = /home/ path_in = /a/../c make_path = c
cur_dir =  path_in = a/./c make_path = /a/c
cur_dir = / path_in = a/./c make_path = /a/c
cur_dir = /home/ path_in = a/./c make_path = /home/a/c
cur_dir =  path_in = /a/./c make_path = /a/c
cur_dir = / path_in = /a/./c make_path = /a/c
cur_dir = /home/ path_in = /a/./c make_path = /a/c
```
There are some outputs like this one: `cur_dir = /home/ path_in = /a/../c make_path = c` which don't start in a / which should.

robert-hh commented:
@davehylands Found it. It was this statement:
`            new_path[len_path == 0 ? 1 : len_path] = '\0';`
I could change it to:
`            new_path[len_path == 0 ? ++len_path : len_path] = '\0';`
But for clarity I prefer:
```
            if (len_path == 0) { // backed up to the tip
                new_path[++len_path] = '\0';
            } else {
                new_path[len_path] = '\0';
            }
```
and let the compiler do the optimization.

robert-hh committed:
/extmod/vfs_lfsx.c: fix error when backing up to /

This happened when replacing strrchr.

robert-hh committed:
extmod/vfs_lfs(x).c; Change the declaration of get_token

Attempt to silence Travis complaints about redefinition of get_token.

petrkr commented:
I can confirm this fixes ".." issue.

This is essential fix for our shell() because it acted strange between version "v1.12 on 2019-12-20" to "v1.12-388-g388d419ba on 2020-04-22"

```
MicroPython v1.12 on 2019-12-20; ESP32 module with ESP32
Type "help()" for more information.
>>> uos.getcwd()
'/'
>>> uos.mkdir("test")
>>> uos.chdir("test")
>>> uos.getcwd()
'/test'
>>> uos.chdir("..")
>>> uos.getcwd()
'/'
>>>
```

```
MicroPython v1.12-388-g388d419ba on 2020-04-22; ESP32 module with ESP32
Type "help()" for more information.
>>> 
paste mode; Ctrl-C to cancel, Ctrl-D to finish
=== uos.getcwd()
=== uos.mkdir("test")
=== uos.chdir("test")
=== uos.getcwd()
=== uos.chdir("..")
=== uos.getcwd()
=== 
'/'
'/test'
'/test/..'
>>> 
```

And recompilled with this PR
```
MicroPython v1.12-430-g603017a09-build-octopusLAB on 2020-05-01; ESP32 module with ESP32
Type "help()" for more information.
>>> 
paste mode; Ctrl-C to cancel, Ctrl-D to finish
=== uos.getcwd()
=== uos.mkdir("test")
=== uos.chdir("test")
=== uos.getcwd()
=== uos.chdir("..")
=== uos.getcwd()
'/'
'/test'
'/'
>>> 
```

robert-hh commented:
@tve: I'm still feeling uncomfortable about allocating a buffer with m_malloc in make_path. Even if the usage periods of that memory is short, typically just a few statements, one of these is a call of a lfs action, which may trigger a garbage collection, which in turn may unexpectedly free this buffer. 

So I come back to your suggestion, that the caller of make_path supplies the address of a pointer (`char **new_path` resp. `char *new_path[]`), which will hold the address. And since that one lives on the stack, if will be preserved as root pointer. Obviously that makes the generated code of make_path a little bit longer.

But maybe you or someone else knows that my concerns are wrong.

tve commented:
Rather than do more code reorg I'd wait for Damien to chime in...

robert-hh commented:
Looking at it again, it looks fine the way it is, because the returned pointer is stored in  a local variable of the caller, which lives in the stack. So it should be safe during every call form that scope.

dpgeorge commented:
I think littlefs does actually handle `"."` and `".."`.  For example, `open("foo/.././test", "w")` seems to work already and creates a file `"test"` in the current directory.

What is the use case that this PR is trying to fix?  It could be that only `chdir()` needs to explicitly handle these dot-files, not all functions.

robert-hh commented:
It is definitely chdir that did not respond well. But chdir does not use any of the littlefs internals.
As far as my short trial tells, open, rename, remove, rmdir and listdir work, and stat as well with the fix for stat. So it's chdir only. I had my doubts, whether such a large change is justified for a little "inconvenience". I wonder if the code which handles . and.. in lfs could be used for that purpose.

dpgeorge commented:
> I wonder if the code which handles . and.. in lfs could be used for that purpose.

I doubt it... that code seems quite internal to the littlefs library.

But it should be rather "easy" to make chdir work correctly: parse the argument to this function and apply it to the current directory vstr.  Eg if there's a `./` then do nothing.  If there's a `../` then chop the previous dir.  Else append the component.  And this doesn't require any extra memory allocation, just reuse the existing `cur_dir` vstr.

robert-hh commented:
But unless this code is used for chdir, the cur_dir vstr must not be shortened. The way it works at the monent is, that the code appends chunks to cur_dir, leaving head part of cur_dir unmodified. That way further use of cur_dir still works. But if .. cuts off chunks from the cur_dir vstr, then things will go wrong.

tve commented:
I think you didn't interpret Damien's last comment correctly. He's suggesting to only fix chdir, and you can do that by changing cur_dir directly (since that's the purpose of the function) instead of producing a normalized path in allocated memory and then copying it to the cur_dir.

robert-hh commented:
If that is the intention then it would work. That also bans the use of make_path at all other places, like for instance stat(), ...
And the change in the code is minor. Still the incoming path has to be scanned, backing up is the same work, only attaching a chunk to the path can be done with the vtrs mechanisms.

robert-hh commented:
So I made a variant of make_path - i called it make_cwd, which works on the cur_dir vstr. Actually it contains almost all  the code needed for a revised chdir. Only the test for existence and appending / is to be added. BUT, the test for existence is the problem. If the new assembled path for CWD does not exist, it leaves cur_dir in an invalid state. To avoid that, either the old value of cur_dir has to be kept, or cur_dir is reset to a known safe state like root. Below is the make_cwd code:
```
#define CWD_LEN (self->cur_dir.len)
#define CWD_BUF (self->cur_dir.buf)

const char *MP_VFS_LFSx(make_cwd)(MP_OBJ_VFS_LFSx * self, mp_obj_t path_in) {
    const char *path = mp_obj_str_get_str(path_in);
    size_t len_token;

    // If cwd was empty (unlikely) or path starts with /, the new cwd start with "/"
    // otherwise keep it, but cut off a trailing '/', which is added by chdir
    if (path[0] == '/' || CWD_BUF[0] == '\0') {
        vstr_reset(&self->cur_dir);
        vstr_add_byte(&self->cur_dir, '/');
    } else { // remove trailing /, if any
        if (CWD_LEN > 1 && CWD_BUF[CWD_LEN - 1] == '/') {
            CWD_LEN--;
        }
    }

    for (char *token = MP_VFS_LFSx(get_token)(path, '/', &len_token); token != NULL;
         token = MP_VFS_LFSx(get_token)(NULL, '/', &len_token)) {
        if (len_token == 2 && token[0] == '.' && token[1] == '.') { // double dot, skip back cwd
            for (--CWD_LEN; CWD_LEN > 0 && CWD_BUF[CWD_LEN] != '/'; CWD_LEN--) ; // skip back 
            if (CWD_LEN == 0) { // backed up to the tip
                CWD_LEN++;
            }
        } else if (len_token != 1 || token[0] != '.') {
            if (CWD_LEN > 1) { // not at the start
                vstr_add_byte(&self->cur_dir, '/');
            }
            vstr_add_strn(&self->cur_dir, token, len_token);
        }
    }
    return vstr_null_terminated_str(&self->cur_dir);
}
```

tve commented:
Haha! Sometimes seemingly simple problems turn out not to be so simple! It's all about perseverance...

WRT to the problem at hand... :-) Suppose you have directories `/a/b` and `cur_dir=="/a"`, now, what are the allowable results of `chdir foo/../b`? I would argue that "error: 'foo' not found" is a valid result. This could lead to an implementation of chdir where instead of appending and normalizing you step through the argument's path components one by one and adjust `cur_dir` with existence check. Basically it turns `chdir(path)` into `for p in path.split("/"): chdir(p)` (ignoring abs vs rel paths for a sec).

The issue I see with that is the state of `cur_dir` after an error. The command `chdir b/foo` would leave `cur_dir=="/a/b"` because after moving to `b` it would find that `foo` doesn't exist and return an error, whereas one might argue that chdir should be atomic and not change `cur_dir` if there is an error. Hmmm, thinking about it a bit, the atomicity argument is quite valid, 'cause it leaves things in an unpredictable state on error otherwise, which essentially means that any sane caller of chdir would have to save (i.e. allocation) `cur_dir` in order to restore it on error.

robert-hh commented:
@tve: The way it works at the moment is, that during normalization it ignores non-existing components and just looks, whether the final path exists. That seems  acceptable to me.
Give the way the whole PR developed, I see two options:
a) use the mechanism as proposed, and maybe use a different mechanism for allocation of the memory for the new path.  That would also allow a simple fix for rename, which is still wrong.
b) close this PR and start over with a variant, where the normalization would only happen within the chdir fucntion. chdir would first use the former make_path to check the resulting path for existence, and the normalize the cur_dir path in situ. Since that path can only get shorter, no additional memory is required.

dpgeorge commented:
> close this PR and start over with a variant, where the normalization would only happen within the chdir fucntion. chdir would first use the former make_path to check the resulting path for existence, and the normalize the cur_dir path in situ.

I would go for this option because it should be relatively simple, and it does not require allocating extra memory anywhere (above what is already possibly allocated by make_path).  Note that the existing chdir implementation already verifies the new path (using make_path and stat).  So it looks like the only thing to do is normalise the new `cur_dir` which can be done in-place in this variable.

robert-hh commented:
I have done that coding already & I am testing & polishing it. Will come until tomorrow.
